sph/include/sph/positions.hpp

```diff
@@ -113,8 +113,12 @@ void computePositionsHost(size_t startIndex, size_t endIndex, Dataset& d, const
         util::tie(d.x_m1[i], d.y_m1[i], d.z_m1[i]) = util::tie(X_m1[0], X_m1[1], X_m1[2]);
         util::tie(d.vx[i], d.vy[i], d.vz[i])       = util::tie(V[0], V[1], V[2]);
 
-        T cv = haveMui ? idealGasCv(d.mui[i]) : constCv;
-        d.temp[i] += energyUpdate(dt, dt_m1, d.du[i], d.du_m1[i]) / cv;
+        T cv    = haveMui ? idealGasCv(d.mui[i]) : constCv;
+        T u_old = cv * d.temp[i];
+        T u_new = u_old + energyUpdate(dt, dt_m1, d.du[i], d.du_m1[i]);
+        // To prevent u < 0 (when cooling with GRACKLE is active)
+        if (u_new < 0.) { u_new = u_old * std::exp(u_new * dt / u_old); }
+        d.temp[i]  = u_new / cv;
```

Comment:
This will not prevent negative temperatures for the GPU version. You could do this:
```c++
T u_new = energyUpdate(u_old, dt, dt_m1, d.du[i], d.du_m1[i]);
```
 `energyUpdate` is called on the GPU as well. But you'll have to correct the call in `positions_gpu.cu`